theano/gpuarray/roi_pool.c

```diff
@@ -0,0 +1,220 @@
+#section kernels
+
+#kernel ROIPoolGPUFwd_kernel : size, *, size, size, size, size, size, size, *, *, * :
+
+KERNEL void ROIPoolGPUFwd_kernel(
+    const ga_size nloops, GLOBAL_MEM const DTYPE_i0 * bottom_data,
+    const DTYPE_i0 spatial_scale, const ga_size channels, const ga_size height,
+    const ga_size width, const ga_size pooled_height, const ga_size pooled_width,
+    const DTYPE_i0* bottom_rois, DTYPE_i0* top_data, DTYPE_i0* argmax_data) {
+    
+    for (ga_size index = 0; index < nloops; ++index) {
+
+        ga_size pw = index % pooled_width;
+        ga_size ph = (index / pooled_width) % pooled_height;
+        ga_size c = (index / pooled_width / pooled_height) % channels;
+        ga_size n = index / pooled_width / pooled_height / channels;
+        bottom_rois += n * 5;
+        ga_size roi_batch_index = bottom_rois[0];
+        ga_size roi_start_w = round(bottom_rois[1] * spatial_scale);
+        ga_size roi_start_h = round(bottom_rois[2] * spatial_scale);
+        ga_size roi_end_w = round(bottom_rois[3] * spatial_scale);
+        ga_size roi_end_h = round(bottom_rois[4] * spatial_scale);
+        ga_size roi_width = fmax((DTYPE_i0) (roi_end_w - roi_start_w + 1) * 1.0, 1.0);
+        ga_size roi_height = fmax((DTYPE_i0) (roi_end_h - roi_start_h + 1) * 1.0, 1.0);
+        DTYPE_i0 bin_size_h = (DTYPE_i0)(roi_height)
+                           / (DTYPE_i0)(pooled_height);
+        DTYPE_i0 bin_size_w = (DTYPE_i0)(roi_width)
+                           / (DTYPE_i0)(pooled_width);
+
+        ga_size hstart = (ga_size)(floor((DTYPE_i0)ph * bin_size_h));
+        ga_size wstart = (ga_size)(floor((DTYPE_i0)pw * bin_size_w));
+        ga_size hend = (ga_size)(ceil((DTYPE_i0)(ph + 1)
+                                         * bin_size_h));
+        ga_size wend = (ga_size)(ceil((DTYPE_i0)(pw + 1)
+                                         * bin_size_w));
+
```

Comment:
You need to subtract/add a small epsilon value within the floor/ceiling function to ensure that boundary cases do not fail due to numerical precision.